# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/quota/InitializationUtils.h
# Commit: 84592335a026
# Full Hash: 84592335a0265529a21335ca9cdc3b1f15da7d82
# Author: Jan Varga <jvarga@mozilla.com>
# Date: 2024-10-15 04:16:11
# Regressor Bug: 1866402
# File Overlap Count: 1
# Description:
#   Bug 1866402 - Reduce code duplication in QuotaManager::OpenStorageDirectory and QuotaManager::OpenClientDirectory; r=dom-storage-reviewers,jari
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D223460
# ==============================================================================

diff -r 46a5babaafd2 -r 84592335a026 dom/quota/InitializationUtils.h
--- /dev/null	Thu Jan 01 00:00:00 1970 +0000
+++ b/dom/quota/InitializationUtils.h	Mon Oct 14 10:06:14 2024 +0000
@@ -0,0 +1,69 @@
+/* -*- Mode: C++; tab-width: 8; indent-tabs-mode: nil; c-basic-offset: 2 -*- */
+/* vim: set ts=8 sts=2 et sw=2 tw=80: */
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#ifndef DOM_QUOTA_INITIALIZATIONUTILS_H_
+#define DOM_QUOTA_INITIALIZATIONUTILS_H_
+
+#include <functional>
+
+#include "mozilla/MozPromise.h"
+#include "mozilla/RefPtr.h"
+#include "mozilla/dom/quota/DirectoryLock.h"
+#include "mozilla/dom/quota/DirectoryLockInlines.h"
+#include "mozilla/dom/quota/ForwardDecls.h"
+
+namespace mozilla::dom::quota {
+
+template <typename Callable>
+class MaybeInitializeHelper final {
+ public:
+  MaybeInitializeHelper(RefPtr<UniversalDirectoryLock> aDirectoryLock,
+                        Callable&& aCallable)
+      : mDirectoryLock(std::move(aDirectoryLock)),
+        mCallable(std::move(aCallable)) {}
+
+  RefPtr<BoolPromise> operator()(
+      const BoolPromise::ResolveOrRejectValue& aValue) {
+    if (aValue.IsReject()) {
+      SafeDropDirectoryLockIfNotDropped(mDirectoryLock);
+
+      return BoolPromise::CreateAndReject(aValue.RejectValue(), __func__);
+    }
+
+    if (!mDirectoryLock) {
+      return BoolPromise::CreateAndResolve(true, __func__);
+    }
+
+    return mCallable(std::move(mDirectoryLock));
+  }
+
+ private:
+  RefPtr<UniversalDirectoryLock> mDirectoryLock;
+  Callable mCallable;
+};
+
+template <typename Callable>
+auto MaybeInitialize(RefPtr<UniversalDirectoryLock> aDirectoryLock,
+                     Callable&& aCallable) {
+  return MaybeInitializeHelper(std::move(aDirectoryLock),
+                               std::forward<Callable>(aCallable));
+}
+
+auto MaybeInitialize(RefPtr<UniversalDirectoryLock> aDirectoryLock,
+                     RefPtr<QuotaManager> aQuotaManager,
+                     RefPtr<BoolPromise> (QuotaManager::*aMethod)(
+                         RefPtr<UniversalDirectoryLock>)) {
+  return MaybeInitializeHelper(
+      std::move(aDirectoryLock),
+      [quotaManager = std::move(aQuotaManager),
+       method = aMethod](RefPtr<UniversalDirectoryLock> aDirectoryLock) {
+        return (quotaManager->*method)(std::move(aDirectoryLock));
+      });
+}
+
+}  // namespace mozilla::dom::quota
+
+#endif  // DOM_QUOTA_INITIALIZATIONUTILS_H_
